listenbrainz_spark/recommendations/recording/user_similarity.py

```diff
@@ -0,0 +1,98 @@
+import sys
+import math
+
+from flask import current_app
+from pyspark.mllib.linalg.distributed import CoordinateMatrix, MatrixEntry
+from pyspark.ml.stat import Correlation
+from pyspark.sql.functions import struct, collect_list
+
+import listenbrainz_spark
+from listenbrainz_spark import SparkSessionNotInitializedException, utils, path
+from listenbrainz_spark.exceptions import PathNotFoundException, FileNotFetchedException
+
+
+def create_messages(similar_users_df):
+    itr = similar_users_df.toLocalIterator()
+    message = {}
+    for row in itr:
+        message[row.user_name] = {user.other_user_name: user.similarity for user in row.similar_users}
+    yield {
+        'type': 'similar_users',
+        'data': message
+    }
+
+
+def threshold_similar_users(matrix, threshold):
+    rows, cols = matrix.shape
+    similar_users = list()
+
+    max_similarity = None
+    min_similarity = None
+    for x in range(rows):
+        for y in range(cols):
+            value = float(matrix[x, y])
+            if x == y or math.isnan(value):
+                continue
+
+            if max_similarity is None:
+                max_similarity = value
+                min_similarity = value
+
+            max_similarity = max(value, max_similarity)
+            min_similarity = min(value, min_similarity)
+
+    similarity_range = max_similarity - min_similarity
+    for x in range(rows):
+        for y in range(cols):
+            if x == y:
+                continue
+
+            similarity = (float(matrix[x, y]) - min_similarity) / similarity_range
+            if similarity >= threshold:
+                similar_users.append((x, y, similarity))
+
+    return similar_users
+
+
+def main(threshold):
+
+    current_app.logger.info('Start generating similar user matrix')
+    try:
+        listenbrainz_spark.init_spark_session('User Similarity')
+    except SparkSessionNotInitializedException as err:
+        current_app.logger.error(str(err), exc_info=True)
+        raise
+
+    try:
+        playcounts_df = utils.read_files_from_HDFS(path.USER_SIMILARITY_PLAYCOUNTS_DATAFRAME)
+        users_df = utils.read_files_from_HDFS(path.USER_SIMILARITY_USERS_DATAFRAME)
+    except PathNotFoundException as err:
+        current_app.logger.error(str(err), exc_info=True)
+        raise
+    except FileNotFetchedException as err:
+        current_app.logger.error(str(err), exc_info=True)
+        raise
+
+    tuple_mapped_rdd = playcounts_df.rdd.map(lambda x: MatrixEntry(x["recording_id"], x["user_id"], x["count"]))
```

Comment:
style thing again, but comments in this and the next few lines would greatly enhance readability in my opinion. It's not obvious at the outset what a co-ordinate matrix or a indexed-row-matrix or vectors_mapped_rdd is

Reply:
I have added some description in the docstring.